File: p02ab.clj
;;;;    Advent of Code 2023 - Day 2 part One and Two
;;;;    https://adventofcode.com/2023/day/2
;;;;
;;;;    Solutions in Clojure
;;;;    (Ter leering ende vermaeck...)
;;;;
;;;;    The sum of possible game ID is:       2545
;;;;    The sum of the power of all sets is: 78111
;;;;
;;;;    (cl) by Arno Jacobs, 2023-12-04
;;;;    

(ns p02ab
  (:require [clojure.string :as str]))


;;;; First - read the data-set
(defn get-lines [file]
  (->> file
       slurp
       (str/split-lines)))

(def data-set (get-lines "./data/inputDay02_2023.txt"))

(def max-red 12)
(def max-green 13)
(def max-blue 14)


;;; Part 1

;;; Will return 0 if string does NOT start with a digit 
(defn parse-int [s]
  (Integer. (re-find #"[0-9]*" (apply str (cons \0 s)))))

;;; Check if 'long' string starts with 'short' string
(defn equal-short-start [short long]
  (= (compare short (apply str (take (count short) long))) 0))

;;; Get the maximum colour values per Game line
(defn get-maximum [colour parts maximum-value]
  (let [value (parse-int (first parts))]
  (if (< (count parts) 2)
     maximum-value
     (if (and (equal-short-start colour (nth parts 1))
              (< maximum-value value)) 
        (get-maximum colour (drop 2 parts) value)
        (get-maximum colour (drop 2 parts) maximum-value)))))
     
(defn maximum-RGB [line]
  (let [parts (str/split line #" ")]
    (list (parse-int (nth parts 1))
          (get-maximum "red" parts 0)
          (get-maximum "green" parts 0)
          (get-maximum "blue" parts 0))))

(defn possible-game [game] 
  (and (>= max-red (nth game 1))
       (>= max-green (nth game 2))
       (>= max-blue (nth game 3))))
  
(defn summed-possible-games [games]
  (let [maximum-RGBs (map maximum-RGB games)]
    (->> maximum-RGBs
         (filter possible-game)
         (map first)
         (reduce +))))

;;; Part 2

(defn find-power [game]
  (->> game 
       maximum-RGB
       rest
       (reduce *)))

(defn find-powers [games]
  (->> games
       (map find-power)
       (reduce +)))


;;; The 'main' program - - -
(defn program []
  (println "Advent of Code 2023 - day 1  (Clojure)")
  (print   "The sum of possible game ID is:       ")
  (println (summed-possible-games data-set))
  (print   "The sum of the power of all sets is: ")
  (println (find-powers data-set))
  (println "0K.\n"))

;; Run in terminal via: clojure -M p01ab.clj
(program)

File: p02ab.hs
-- Advent of Code 2023 - Day 2 part One and Two
-- Solutions in Haskell
-- (Ter leering ende vermaeck...)
--
-- The sum of possible game ID is:       2545
-- The sum of the power of all sets is: 78111
--
-- (cl) by Arno Jacobs, 2023-12-02

-- 
module AoC2023d02ab where

import Data.Char

-- Some initials
filename :: String
filename = "data/inputDay02_2023.txt"

type RGBvalues = (Int,Int,Int) -- rgb values


noCubes = (0,0,0) :: RGBvalues

( maxRed, maxGreen, maxBlue ) = (12,13,14) :: RGBvalues

cMinus = '-' :: Char
cSpace = ' ' :: Char

-- I didn't go for the 'splitOneOf' function
-- splitOneOf ":,;" game
-- For example resulting in: ["Game 1"," 20 green"," 3 red"," 2 blue"," 9 red",...
--

extractInt :: String -> (Int,String)
extractInt []       = (0,[])
extractInt (x:xs)   | x == cMinus   = (i,rs)
                    | isDigit x     = (i,rs)
                    | otherwise     = extractInt xs
    where
        i  = read (x:(takeWhile isDigit xs))
        rs = dropWhile isDigit xs 

isLowerCase :: Char -> Bool
isLowerCase c = (c >= 'a') && (c <= 'z')

extractColour :: String -> (String,String)
extractColour []    = ([],[])
extractColour xs    = (takeWhile isLowerCase start, dropWhile isLowerCase start)
    where
        start   = dropWhile (not . isLowerCase) xs

-- Works from start of game info
getGame :: String -> (Int,String)
getGame gameInfo = ( gameNumber, tail remainingLine)
    where
        (gameNumber,remainingLine) = extractInt gameInfo

-- part 1
-- summing all colour info
--
workGame :: String -> (Int,RGBvalues)
workGame game = (gameNumber, workColours colourInfo noCubes)
    where
        (gameNumber,colourInfo) = getGame game

-- The instructions are not clear to me
-- My interpretation is to sum all colours withing one game
-- But, the task is the maximum per set
--
-- Nice is that this part is VERY USABLE for part 2 of the challenge
--
workColours :: String -> RGBvalues -> RGBvalues
workColours []         cubesCount = cubesCount
workColours colourInfo (r,g,b)  
    | colour == "red"   = workColours remainingInfo2 ( (max r cubes), g, b )
    | colour == "green" = workColours remainingInfo2 ( r, (max g cubes), b )
    | colour == "blue"  = workColours remainingInfo2 ( r, g, (max b cubes) )
    | otherwise         = error "Could not extract colour!"
        where
            (cubes,remainingInfo1)  = extractInt colourInfo
            (colour,remainingInfo2) = extractColour remainingInfo1

possibleGame :: (Int,RGBvalues) -> Bool
possibleGame (_,(r,g,b)) = (r <= maxRed) && (g <= maxGreen) && (b <= maxBlue)

workGames :: [String] -> Int
workGames games = sumGameNumbers
    where
        countedCubes    = map workGame games
        filterGames     = filter possibleGame countedCubes
        sumGameNumbers  = sum $ map fst filterGames

-- part 2 
--
findPower :: String -> Int
findPower game = r * g * b
    where
        ( _, colourInfo )   = getGame game
        ( r, g, b )         = workColours colourInfo noCubes

findPowers :: [String] -> Int
findPowers = sum . map findPower

main :: IO ()
main = do   putStrLn "Advent of Code 2023 - day 2  (Haskell)"
            games <- lines <$> readFile filename
            putStr   "The sum of possible game ID is:       "
            print $ workGames games
            putStr   "The sum of the power of all sets is: "
            print $ findPowers games
            putStrLn "0K.\n"

